### DISCUSSION TITLE
Can't query Json field

### DISCUSSION THREAD

**Discussion by alper** [diesel-rs/diesel#3008] [Q&A] [answered]
Hi, 

I have `models.rs` like this:

```rust
use diesel::sql_types::Json;

#[derive(Queryable)]
pub struct GMapsLocation {
    pub id: i32,
    pub place_id: String,
    pub data: Json,
}
```

And trying to do this:
```rust
    let results: Vec<GMapsLocation> = gmaps_locations
        .select((id, place_id, data))
        .load(&connection)
        .expect("Erorr loading locations");
```

And then I get this: 
```
src/main.rs:18:10
     |
18   |         .load::<GMapsLocation>(&connection)
     |          ^^^^ the trait `Queryable<diesel::sql_types::Json, _>` is not implemented for `diesel::sql_types::Json`
     |
     = note: required because of the requirements on the impl of `Queryable<(Integer, Text, diesel::sql_types::Json), _>` for `(i32, std::string::String, diesel::sql_types::Json)`
     = note: 1 redundant requirement hidden
     = note: required because of the requirements on the impl of `Queryable<(Integer, Text, diesel::sql_types::Json), _>` for `foursquare_mapper::models::GMapsLocation`
     = note: required because of the requirements on the impl of `LoadQuery<_, foursquare_mapper::models::GMapsLocation>` for `SelectStatement<table, query_builder::select_clause::SelectClause<(foursquare_mapper::schema::gmaps_locations::id, foursquare_mapper::schema::gmaps_locations::place_id, foursquare_mapper::schema::gmaps_locations::data)>>`
note: required by a bound in `load`
```

No idea what is happening and I can't find an example of how to query Json anywhere in the docs. This also shows how to insert data: https://docs.diesel.rs/diesel/pg/types/sql_types/struct.Jsonb.html

**Comment by alper** [comment]
Inserts are working with this setup though as per the example on the Jsonb page.

**Comment by alper** [comment, marked as answer]
OK. I found out what the issue was.

Just like the input value [from the documentation][1] is a `serde_json::Value`, the return value from the query is also a `serde_json::Value`. I have no clue what `diesel::sql_types::Json` is for then.

Also, we don't get the struct back that's in models but a vector of tuples instead. (Why have a model?)

So the working code is:
```rust
let results = gmaps_locations
    .select((id, place_id, data))
    .load::<(i32, String, serde_json::Value)>(&connection)
    .expect("Erorr loading locations");
```


  [1]: https://docs.diesel.rs/diesel/pg/types/sql_types/struct.Jsonb.html

**Reply by weiznich** [reply]
There is no need to use a tuple here. Any struct that implements Queryable the same way will work there. That includes structs that use a serde_json::Value field at the corresponding location.
As for the use of types from diesel::sql_types: Checkout the API documentation of that module.

**Reply by alper** [reply]
But why didn't it work with my original struct? The one that defines the table?

**Reply by weiznich** [reply]
I cannot see how your struct defines a table. That's what the `table!` macro is doing. As general rule: Never use a type from `diesel::sql_types` to hold any value. They are zero sized types so they cannot hold values. If you use the same type in your struct as in your tuple it will work.

**Reply by alper** [reply]
And the Queryable should not have a Json in its definition? Sorry, I've gone through the documentation as best I could.

**Reply by alper** [reply]
OK. I changed my model to:

```rust
pub struct GMapsLocation {
    pub id: i32,
    pub place_id: String,
    pub data: serde_json::Value,
}
```

And now it's working without the tuple.
